Docs: rewrite data source intro

MattDodsonEnglish opened the pull request:
- Include more than just "databases", which was overly narrow
- Try to make doc a "TL;DR" of the essential information
- Link out to plugins, so readers can find data-source extensions

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

MattDodsonEnglish committed:
Docs: rewrite data source intro

- Include more than just "databases", which was overly narrow
- Try to make doc a "TL;DR" of the essential information

tolzhabayev commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,14 @@ weight: 60
-When you have created and configured a data source you are ready to start exploring and visualizing data, either in Explore or in a new Dashboard. A dashboard is composed of [panels]({{< relref "../panels-visualizations/" >}}), each panel contains a set of queries to one or more data sources.
+Grafana has built-in support for a wide range of _data sources_.
+These data sources might be files, services, or databases.
+You can also extend the supported data sources through one of the many [Plugins](/grafana/plugins/) (or build your own).
```
wrong link?
I guess you wanted to link to https://grafana.com/grafana/plugins/

tolzhabayev replied:
"build your own" could be linked to https://grafana.com/docs/grafana/latest/developers/plugins/

MattDodsonEnglish replied:
Good idea about adding the link to the build instructions. https://github.com/grafana/grafana/pull/61639/commits/b3e48d45922cf3331c71ef0c8fe6251901d59391

tolzhabayev commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,14 @@ weight: 60
-When you have created and configured a data source you are ready to start exploring and visualizing data, either in Explore or in a new Dashboard. A dashboard is composed of [panels]({{< relref "../panels-visualizations/" >}}), each panel contains a set of queries to one or more data sources.
+Grafana has built-in support for a wide range of _data sources_.
+These data sources might be files, services, or databases.
+You can also extend the supported data sources through one of the many [Plugins](/grafana/plugins/) (or build your own).
```
Not sure how we usually talk about them - `supported` vs `built-in`

MattDodsonEnglish replied:
Yeah, that might be a little bit of language from k6 docs. I wanted to avoid jargon like "native" or "out of the box".

I think we can avoid this issue entirely by just using _support_ as a verb.

> Grafana supports a wide range of _data sources_.

Shorter sentence, too :-) .

MattDodsonEnglish replied:
It's also consistent, since it's the exact language of https://grafana.com/tutorials/build-a-data-source-plugin/

MattDodsonEnglish replied:
[54e3654](https://github.com/grafana/grafana/pull/61639/commits/54e3654f2224c8491ad3cc6a031520d8f0a950c2)

MattDodsonEnglish committed:
docs, link data source to build plugin task

MattDodsonEnglish committed:
Reduce jargon and sentence length

MattDodsonEnglish committed:
Link out to different features

MattDodsonEnglish committed:
Data source, docs turn examples into list

And trim language

MattDodsonEnglish committed:
Better phrasing to introduce query editor

Old way implied that Grafana provides all query editors,
which isn't true of custom plugins.

Eve832 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,17 @@ weight: 60
+Each data source comes with a _query editor_, which formulates custom queries according to the source's particular structure.
+After you add a data source, you can use it as an input for many operations:
+- Query the data with [Explore]({{< relref "../explore" >}}).
+- Visualize it in in [panels]({{< relref "../panels-visualizations" >}}).
```
```suggestion
- Visualize it in [panels]({{< relref "../panels-visualizations" >}}).
```

MattDodsonEnglish committed:
use relref for in-repo anchor

Eve832 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,17 @@ weight: 60
-When you have created and configured a data source you are ready to start exploring and visualizing data, either in Explore or in a new Dashboard. A dashboard is composed of [panels]({{< relref "../panels-visualizations/" >}}), each panel contains a set of queries to one or more data sources.
+Grafana supports a wide range of _data sources_.
+These data sources might be files, services, or databases.
+You can also extend the supported data sources through one of the many [Plugins](/grafana/plugins/) (or [build your own](/docs/grafana/latest/developers/plugins/)).
```
So this is a little nuanced, but you can't leverage a data source without a plugin. This makes it seem that plugins do something "extra" to data sources. 

Perhaps:
You can enable data sources by installing and configuring their associated Grafana plugins. For a list of available plugins, see [](). Alternatively, you can create your own, custom plugin. To learn more refer to []()

(I have a personal vendetta against unintroduced hyperlinks—feel free to ignore that)

MattDodsonEnglish replied:
Hmm maybe I misunderstand. I thought Grafana had two models to use data sources:

- Built-in support. These ship with the Grafana binary, and they do not need to be installed.
- Extended through plugins. These have to be added from the available plugin list, or custom built.

Am I off here? I see that in some instances, the Grafana site says "plugin" to mean built-in data source.

> Loki Data Source - Native Plugin
> This is a built in data source that allows you to connect to the Loki logging service. [^1]

In my opinion, Grafana should avoid calling things that ship with the base binary a "plugin". I worry that this could corrode associations of meaning in the reader's heads, since, as far as I'm aware, plugin typically means an optional component to add to the system. It implies some customizability and tells people who don't want to use this data source that they can keep their binary footprint lighter by not installing it.

To call Loki a plugin implies that it could be plugged out, but that's not really true―it'd be more like rewiring.

I think saying things like "supported data source" or "built-in support" reduces cognitive reader load. To use _plugin_ to mean "outside of the base binary" creates a clear distinction. This is all to say that, if I haven't missed something, I'd prefer to try to reduce terms like "native plugin" in other places rather than introduce them here.

Does that make sense?

[^1]: https://grafana.com/grafana/plugins/loki/

MattDodsonEnglish replied:
> This makes it seem that plugins do something "extra" to data sources.

I see your point about this sentence, though. What about this?

```diff
- You can also extend the supported data sources through one of the many [Plugins]
+ You can also add more data sources with Plugins
```

@Eve832 what do you think?  (ignoring the hyperlink―that's not a big deal to me).

Eve832 replied:
There are only like 13 data sources that we natively support (fewer in Cloud I think). There are 80+ data sources. All of those not supported require a plugin

MattDodsonEnglish replied:
I just pushed that text to make the review more streamlined:
https://github.com/grafana/grafana/pull/61639/commits/4bd9fb7535d031e67b546716cc3e9e2d5f36a3d3

Eve832 replied:
Yeah, I this is def more clear

Eve832 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,17 @@ weight: 60
+After you add a data source, you can use it as an input for many operations:
+- Query the data with [Explore]({{< relref "../explore" >}}).
+- Visualize it in in [panels]({{< relref "../panels-visualizations" >}}).
+- Make criteria for [alerts]({{< relref "../alerting" >}}).
```
```suggestion
- Create rules for [alerts]({{< relref "../alerting" >}}).
```

Eve832 reviewed (changes requested):
Another thing to consider is that when we talk about data sources, we're talking about a very specific connection type in the ecosystem of Grafana. Data sources are not the same as "sources of data." This is why devs have noted them as "datasources" in the past. 
A subtle, important, and often very confusing nuance.

MattDodsonEnglish committed:
remove double "in"

Co-authored-by: Eve Meelan <81647476+Eve832@users.noreply.github.com>

MattDodsonEnglish committed:
Simplify list

Co-authored-by: Eve Meelan <81647476+Eve832@users.noreply.github.com>

MattDodsonEnglish commented:
>  Data sources are not the same as "sources of data."

What's the difference? It seems to me like a data source is any data structure, whether static or served, that can be queried from a Grafana query editor.

Eve832 commented:
> > Data sources are not the same as "sources of data."
> 
> What's the difference? It seems to me like a data source is any data structure, whether static or served, that can be queried from a Grafana query editor.

Data sources are one grafana connection tool tailored for a service. Integrations are another tool. Some sources of data have both. They do different things.

MattDodsonEnglish commented:
> Data sources are one grafana connection tool tailored for a service. Integrations are another tool. Some sources of data have both. They do different things.

Perhaps the difference between integration and data source is a subject for another doc, somewhere Grafana Cloud side? In this particular place, I just want to give the essential facts of what a data source is, and I think to say "data source is not a source of data" could be a bit confusing.

MattDodsonEnglish committed:
Restructure intro, clarify plugins and their URLs

MattDodsonEnglish committed:
Add and configure

It's sometimes both.

Eve832 commented:
> > Data sources are one grafana connection tool tailored for a service. Integrations are another tool. Some sources of data have both. They do different things.
> 
> Perhaps the difference between integration and data source is a subject for another doc, somewhere Grafana Cloud side? In this particular place, I just want to give the essential facts of what a data source is, and I think to say "data source is not a source of data" could be a bit confusing.

This content does exist in Cloud, but integrations will likely come to OSS in the coming months. The term "data source" has a very specific meaning to OSS users. We're trying to move away from this by terming all connections, "Connections" within the Connections Console, but it's been very difficult because they really are a specific "thing." 

This is more of a statement of connections history, rather than a recommendation.

MattDodsonEnglish commented:
> This content does exist in Cloud, but integrations will likely come to OSS in the coming months. The term "data source" has a very specific meaning to OSS users. We're trying to move away from this by terming all connections, "Connections" within the Connections Console, but it's been very difficult because they really are a specific "thing."

> This is more of a statement of connections history, rather than a recommendation.

Ah I see. This is good info, and I didn't know anything about that. So, thanks!

Still, I'm happy to accept that this data-source doc is marginally better today, and we can kick the connections can down the road. Truthfully, I'm too cowardly to get into data connections today. Maybe I'll find more courage later :-). 

If you're happy to merge, it's fine with me. I think this third incarnation of the intro is the clearest and most coherent. Thanks to all reviewers.

chri2547 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,21 @@ weight: 60
+- Visualize it in [panels]({{< relref "../panels-visualizations" >}}).
+- Create rules for [alerts]({{< relref "../alerting" >}}).
+
+This documentation goes over how to manage data sources in general,
```
```suggestion
This documentation describes how to manage data sources in general,
```

chri2547 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,21 @@ weight: 60
+This documentation goes over how to manage data sources in general,
+and how to configure or query the built-in data sources.
+For other data sources, refer to the list of [datasource plugins](/grafana/plugins/).
+To make a custom plugin, refer to [Build a plugin]({{< relref "../developers/plugins/" >}}).
```
```suggestion
To develop a custom plugin, refer to [Build a plugin]({{< relref "../developers/plugins/" >}}).
```

chri2547 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,21 @@ weight: 60
-You can also query data sources without building a dashboard by using the [Explore]({{< relref "../explore/" >}}) feature.
+Grafana comes with built-in support for many _data sources_.
+If you need other data sources, you can also install one of the many data source plugins.
+If the plugin you need doesn't exist, you can also create your own custom plugin.
```
```suggestion
If the plugin you need doesn't exist, you can develop a custom plugin.
```

chri2547 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,21 @@ weight: 60
+If the plugin you need doesn't exist, you can also create your own custom plugin.
+
+Each data source comes with a _query editor_,
+which formulates custom queries according to the source's particular structure.
```
```suggestion
which formulates custom queries according to the source's structure.
```

chri2547 commented on docs/sources/datasources/_index.md:
```diff
@@ -8,12 +8,21 @@ weight: 60
+
+Each data source comes with a _query editor_,
+which formulates custom queries according to the source's particular structure.
+After you add and configure a data source, you can use it as an input for many operations:
```
```suggestion
After you add and configure a data source, you can use it as an input for many operations, including:
```

chri2547 reviewed (commented):
Nice, thanks for the contribution! A couple of suggestions.

MattDodsonEnglish committed:
Apply suggestions from code review

Co-authored-by: Christopher Moyer <35463610+chri2547@users.noreply.github.com>

chri2547 committed:
makes prettier
